### PR TITLE
[NCCL] Overwrite NCCL Stream in AllReduce

### PR DESCRIPTION
**Pull request by osalpekar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43675 Test for NCCL Stream Overwrite
* #41356 [NCCL] Expose NCCL-specific collective options to Python
* #41276 [NCCL] Adding CUDA Stream Overwrite Options for Allgather, Reduce, ReduceScatter, and Broadcast
* **#41275 [NCCL] Overwrite NCCL Stream in AllReduce**

Allows user to pass a vector of CUDA stream objects through AllReduce Options, and use those stream instead of default streams to enqueue GPU operations.

Differential Revision: [D22242878](https://our.internmc.facebook.com/intern/diff/D22242878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22242878/)!

### CODE REVIEW

**Comment by jiayisuse** [comment]
Seems we didn't update Python interface. Let's bind NCCLAllreduceOptions with python?

**Comment by osalpekar** [comment]
> Seems we didn't update Python interface. Let's bind NCCLAllreduceOptions with python?

Added Python bindings in https://github.com/pytorch/pytorch/pull/41356

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -801,15 +801,25 @@ template <typename Fn, typename PreProcess, typename PostProcess>
...
     PostProcess post) {
   const auto devices = getDeviceList(inputs);
   const auto key = getKeyFromDevices(devices);
   auto& ncclComms = getNCCLComm(key, devices);
 
+  auto& streams = cudaStreams.empty() ? ncclStreams_[key] : cudaStreams;
+
+  // Check the size of the streams vector (this is necessary when we're using
+  // the user-supplied CUDA streams).
+  if (streams.size() != ncclStreams_[key].size()) {
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:816]: Probably want to compare with `devices.size()` here.

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -801,15 +801,25 @@ template <typename Fn, typename PreProcess, typename PostProcess>
 std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
     std::vector<at::Tensor>& inputs,
     std::vector<at::Tensor>& outputs,
+    std::vector<at::cuda::CUDAStream> cudaStreams,
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:804]: pass as a const ref?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -864,6 +874,21 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
   return work;
 }
 
+template <typename Fn>
+std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+    std::vector<at::Tensor>& inputs,
+    std::vector<at::Tensor>& outputs,
+    std::vector<at::cuda::CUDAStream> cudaStreams,
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:881]: const ref?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 4]
```diff
@@ -872,19 +897,33 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
   return collective(
       inputs,
       outputs,
+      std::vector<at::cuda::CUDAStream>(),
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:900]: nit: I guess `{}` would work here?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 5]
```diff
@@ -872,19 +897,33 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
...
 
+template <typename Fn, typename PreProcess, typename PostProcess>
+std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+    std::vector<at::Tensor>& inputs,
+    std::vector<at::Tensor>& outputs,
+    Fn fn,
+    PreProcess pre,
+    PostProcess post) {
+  return collective(
+      inputs, outputs, std::vector<at::cuda::CUDAStream>(), fn, pre, post);
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:914]: nit: `{}` here as well?

File: torch/lib/c10d/Types.hpp [hunk 6]
```diff
@@ -25,8 +29,12 @@ struct BroadcastOptions {
 };
 
 struct AllreduceOptions {
+  virtual ~AllreduceOptions() = default;
   ReduceOp reduceOp = ReduceOp::SUM;
   std::chrono::milliseconds timeout = kUnsetTimeout;
+/* #ifdef USE_C10D_NCCL */
+/*   std::vector<at::cuda::CUDAStream> cudaStreams = {}; */
+/* #endif */
```
Comment (pritamdamania87) [inline torch/lib/c10d/Types.hpp:37]: Why is the entire block commented out here?

**Review by pritamdamania87 — COMMENTED** [review]
Could you add some details to the PR summary or link to an existing gh issue about the use case for this feature?

It looks like we only add these options in CPP for now, is there a plan to expose these options to python?

**Pull request closed without merging**